### ISSUE TITLE
GUI : Capacity on Demand Features

### ISSUE THREAD

**Issue by derick-montague** [ibm-openbmc/dev#1538]
## SMEs
**BMC**: ~~Ratan Gupta, Justin King, @gtmills~~ @raviteja-b 
**PHYP**: Paul Olson
**RAS**: ~~Justin Thaler, George Ahrens~~, Kanisha Patel
### Summary
Capacity on Demand is a pay-for-use model that allows mid-range and high-end customers to use capacity flexibly by activating and deactivating processor cores and memory units on their server as their business needs change (i.e. during workload peaks). The CoD function controls the number of cores and memory units that the customer can actually use as per the license. The only CoD functions that apply to low-end servers (e.g. Rainier boxes) are the PowerVM (virtualization) functions. 
PowerVM allows customers to consolidate multiple workloads onto fewer systems, increasing server utilization and reducing costs. 

**How is this different from FCO?** 
CoD allows reducing the functional processor cores only up to a limit. To reduce below that number, FCO needs to be used to override the licensed number of cores.

### User Story
- As an admin, I want to activate (through licensing) processors and memory resources within my system so that I can meet my current environment needs, without taking my system down (non-disruptively). 
- As an eBMC Sys Admin (HMC/no HMC), I want to manage cores in the processors to meet my current environment without the need for buying/shipping new hardware. 

### InVision Prototype
- System off: https://ibm.invisionapp.com/d/main#/console/15347903/319985409/preview
- System on: https://ibm.invisionapp.com/d/main#/console/15347903/319985410/preview

### References
- [Discovery folder](https://ibm.ent.box.com/folder/121382866225)
- [Backend Requirements from Engineering architectural meeting](https://ibm.box.com/s/6ggkxsb1rn0pc3jay8c6sllhdobrv3oq)

### UX Flow / Interaction Requirements
- [ ] If system is off, show the _inline_ info notification "Applying changes to ..." on top of the page. 
- [ ] If system is off, and user submits correct activation code, activation will not work. In this case, show error notification "...Error starting activation"
- [ ] If an invalid activation key is submitted, show error notification "...Invalid activation key"
- [ ] If the machine is `stand-alone` not managed/connected to an HMC, we should hide the navigation for that page. If that is a challenge, the form must be disabled. If disabled, we may need to show a message to the user explaining why.
    - Use sessions to determine if the machine is connected to an HMC using the `ClientID` and the parse the value to verify any have the following value, `HMCID=`

### cURL Commands and Redfish Response
Redfish commands:
```
1. Upload the license
==> curl -k -H "X-Auth-Token: $bmc_token" -H "Content-Type: application/json" -X POST -T <file path of activation code string>  https://${bmc}/redfish/v1/LicenseService/Actions/LicenseService.Install

2. List the licenses
==>curl -k -H "X-Auth-Token: $bmc_token" -X GET https://${bmc}/redfish/v1/LicenseService
==>curl -k -H "X-Auth-Token: $bmc_token" -X GET https://${bmc}/redfish/v1/LicenseService/Licenses
```
### Notes
- [ ] Any known API workarounds
- [ ] Any server-side functionality WIP

#### Browser Tests (Chrome, Firefox, Safari (Mac), Edge (Windows))
- [ ] Tested responsive layout on supported viewport sizes
- [ ] Verified functionality works as expected, e.g. success and failure use cases
- [ ] Resolved any JavaScript errors thrown to the console

#### Accessibility Tests
- [ ] Verified markup is semantic
- [ ] No reported issues in IBM Accessibility Checker (Free open source tool) https://www.ibm.com/able/toolkit/tools/

#### Test Hooks
- [ ] Verify need for `data-test-id` hooks with test team

**Comment by derick-montague** [comment]
A mini-workshop discussion today hosted by Dean Sanner led to a discussion about the need for a list of all licensing keys as part of CoD. I don't believe this is part of ASMI, but is an AIX line item. @gtmills suggested some type of licensing page. I am documenting this information for the discovery phase.

TL;DR
- Will be similar to #1290 but AIX only
- Will be saved as part of licensing server
- Will be available in Redfish
- Will contain more than just the AIX key
    - Not clear how many or what license keys will be stored

**Comment by derick-montague** [comment]
Add this story as a blocker to FW Access Key due to mini workshop noted in the above comment. I have also added a similar comment in that story. https://github.com/ibm-openbmc/dev/issues/1290#issuecomment-797827403

**Comment by derick-montague** [comment]
Updated milestone based on feature sync meeting with Maury and Gunnar on 3/15/21.

**Comment by derick-montague** [comment]
@priyanka-pillai97 - You may want to find a FED to work on this with you. Here is some information from Ratan in the #1841, https://github.com/ibm-openbmc/dev/issues/1841#issuecomment-834257902

**Comment by derick-montague** [comment]
Capacity on Demand is reliant on the Redfish License Service. The [conversation about delivery timeline can be found in this slack thread](https://ibm-systems-power.slack.com/archives/G3WAVQYTH/p1627585772079600)

**Comment by dixsie** [comment]
Capacity on Demand Design Review 8/14/21

- Serial number will be there on computer system schema
- For access key info  create a resource dump
- Processor license (cores license), might be 1 field (id and sequence #), same for memory license
- VET resource id - not sure we will have this
- Vet capabilities each have an individual license
- Vet capability status should be enabled/disabled not true/false — should check with customers
-  can have a 3rd state - unknown (can happen when booting for 1st time, or after factory reset)
- Can use genesis mode for license info for vet capabilities 
- System type (model property) - do we need first 4 digits or whole string? Mock and ASMI shows 4 digits, but model property looks like  ####-###
-  Get with @santoshpuranik  for endpoints

**Comment by dixsie** [comment]
Add license support is in code review https://github.ibm.com/openbmc/bmcweb/pull/206

**Comment by dixsie** [comment]
Backend is available on 31.2 driver, as of today we can develop everything in the mockup except the activation section.

**Comment by mzipse** [comment]
Work is complete.  Closing.
